### PR TITLE
Mobx: ArcGisPortal models

### PR DESCRIPTION
**Pull request by rowanwins**
### What this PR does

This PR adds a new `ArcGisPortalGroup` model for harvesting for ArcGIS portals. It also includes an `ArcGisPortalItemReference` which converts things into other Terria models.

I'm currently experiencing some issues in the jasmine tests with the `ArcGisPortalItemReference` not properly creating the `previousTarget` but it seems to work ok in the browser.

Here's some sample configurations. Note that a couple of them require use within something like NSW Digital Twin where we can get a users token.
````
{
 "catalog": [
    {
      "type": "arcgis-portal-item",
      "url": "https://portal.spatial.nsw.gov.au/portal",
      "itemId": "66fabd8c23074ecc85883e0086419adc",
      "name": "Road Segments"
    },
    {
      "type": "arcgis-portal-group",
      "name": "All open Content",
      "url": "https://portal.spatial.nsw.gov.au/portal"
    },
    {
      "type": "arcgis-portal-group",
      "name": "Users Private Content",
      "url": "https://portal.spatial.nsw.gov.au/portal",
      "searchParams": {
        "q": "access:private and type:'Feature Service"
      }
    },
    {
      "type": "arcgis-portal-group",
      "name": "Users Groups",
      "url": "https://portal.spatial.nsw.gov.au/portal",
      "groupBy": "usersGroups"
    },
    {
      "type": "arcgis-portal-group",
      "name": "Organisations Groups",
      "url": "https://portal.spatial.nsw.gov.au/portal",
      "groupBy": "organisationsGroups",
      "groupSearchParams": {
        "q": "orgid:0123456789ABCDEF"
      }
    }
  ]
}
````

### Checklist

-   [ ] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
-   [ ] I've updated CHANGES.md with what I changed.

### CODE REVIEW

**Review by tephenavies — CHANGES_REQUESTED** [review]

File: lib/Traits/ArcGisPortalItemFormatTraits.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    description: "The ID of this items format.",
+    type: "string"
+  })
+  id?: string;
+
+  @anyTrait({
+    name: "Url Regular Expression",
+    description: "A regular expression that is matched against the items url."
+  })
+  urlRegex?: string | RegExp;
```
[the remark this answers is not in the document]
Comment (rowanwins) [reply lib/Traits/ArcGisPortalItemFormatTraits.ts:18]: For some reason these lines ([here](https://github.com/TerriaJS/terriajs/pull/4496/files#diff-ba7e5f5f11487cf6fd0b2e018c843db4R219) and [here](https://github.com/TerriaJS/terriajs/pull/4496/files#diff-ba7e5f5f11487cf6fd0b2e018c843db4R235)) were giving me massive grief, every string variant I tried of them wouldn't work properly, so I gave up

File: lib/Models/ArcGisPortalCatalogGroup.ts [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+  }
+
+  private getGroups(): CatalogGroup[] {
+    if (this._catalogGroup.groupBy === "none") return [];
+
+    let groups: CatalogGroup[] = [];
+    createUngroupedGroup(this, groups);
+    createGroupsByPortalGroups(this, groups);
+    groups.sort(function(a, b) {
+      if (a.name === undefined || b.name === undefined) return 0;
```
[the remark this answers is not in the document]
Comment (rowanwins) [reply lib/Models/ArcGisPortalCatalogGroup.ts]: addressed

File: lib/Traits/ArcGisPortalItemFormatTraits.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rowanwins) [reply lib/Traits/ArcGisPortalItemFormatTraits.ts:18]: Just to clarify the example further

````
const r = /MapServer\/\d/
const r2 = new RegExp(/MapServer\/\d/)
const r3 = new RegExp('/MapServer\/\d/')

const url = 'https://portal.spatial.nsw.gov.au/server/rest/services/NSW_Features_of_Interest_Category/MapServer/0'

r.test(url)=> true
r2.test(url)=> true
r3.test(url)=> false
````

[the remark this answers is not in the document]
Comment (soyarsauce) [reply lib/Traits/ArcGisPortalItemFormatTraits.ts:18]: Give this a whirl:
`const r4 = new RegExp('MapServer\/\\d')`

```
r4.test(url);
// true
```

**Review by soyarsauce — APPROVED** [review]
Let's merge this & address https://github.com/TerriaJS/terriajs/pull/4496#pullrequestreview-457147754 as part of another ticket

File: lib/Traits/ArcGisPortalItemFormatTraits.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (soyarsauce) [reply lib/Traits/ArcGisPortalItemFormatTraits.ts:18]: https://github.com/TerriaJS/terriajs/issues/4593

**Review by tephenavies — APPROVED** [review]

**Pull request merged**